spyder/plugins/console/widgets/main_widget.py

```diff
@@ -0,0 +1,535 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""
+Main Console widget.
+"""
+
+# pylint: disable=C0103
+# pylint: disable=R0903
+# pylint: disable=R0911
+# pylint: disable=R0201
+
+# Standard library imports
+import logging
+import os
+import os.path as osp
+import sys
+
+# Third party imports
+from qtpy.compat import getopenfilename
+from qtpy.QtCore import Qt, Signal, Slot
+from qtpy.QtWidgets import QInputDialog, QLineEdit, QMenu, QVBoxLayout
+
+# Local imports
+from spyder.api.translations import get_translation
+from spyder.api.widgets import PluginMainWidget
+from spyder.config.base import DEV, get_debug_level
+from spyder.plugins.console.widgets.internalshell import InternalShell
+from spyder.py3compat import to_text_string
+from spyder.utils.misc import (get_error_match, getcwd_or_home,
+                               remove_backslashes)
+from spyder.utils.qthelpers import mimedata2url
+from spyder.widgets.findreplace import FindReplace
+from spyder.widgets.reporterror import SpyderErrorDialog
+
+
+# Localization
+_ = get_translation('spyder')
+logger = logging.getLogger(__name__)
+
+
+# --- Constants
+# ----------------------------------------------------------------------------
+class ConsoleWidgetActions:
+    # Triggers
+    Environment = 'environment_action'
+    ExternalEditor = 'external_editor_action'
+    MaxLineCount = 'max_line_count_action'
+    Quit = 'quit_action'
+    Run = 'run_action'
+    SysPath = 'sys_path_action'
+
+    # Toggles
+    ToggleCodeCompletion = 'toggle_code_completion_action'
+    ToggleWrap = 'toggle_wrap_action'
+
+
+class ConsoleWidgetMenus:
+    InternalSettings = 'internal_settings_submenu'
+
+
+class ConsoleWidgetOptionsMenuSections:
+    Run = 'run_section'
+    Quit = 'quit_section'
+
+
+class ConsoleWidgetInternalSettingsSubMenuSections:
+    Main = 'main'
+
+
+# --- Widgets
+# ----------------------------------------------------------------------------
+class ConsoleWidget(PluginMainWidget):
+    DEFAULT_OPTIONS = {
+        'codecompletion/auto': True,
+        'commands': [],
+        'external_editor/gotoline': '',
+        'external_editor/path': '',
+        'max_line_count': 300,
+        'message': 'Internal console\n\n',
+        'multithreaded': False,
+        'namespace': None,
+        'profile': False,
+        'show_internal_errors': True,
+        'wrap': True,
+        # From appearance
+        'color_theme': 'spyder/dark',
+    }
+
+    # Signals
+    sig_edit_goto_requested = Signal(str, int, str)
+    sig_focus_changed = Signal()
+    sig_refreshed = Signal()
+    sig_show_status_requested = Signal(str)
+    sig_quit_requested = Signal()

```

Comment:
Please add comments to describe what these signals do.

Reply:
Fixed